consensus/src/lib.rs

```diff
@@ -26,28 +52,126 @@ pub enum ProgressResult {
     NonNilPreComitted(ConsensusRound, Hash256, Timestamp),
     NilPreVoted(ConsensusRound, Timestamp),
     NilPreComitted(ConsensusRound, Timestamp),
-    Finalized(Timestamp),
+    Finalized(Hash256, Timestamp),
+}
+
+pub struct ConsensusMessageFilter {
+    /// Note that it is even DESIRABLE to use a synchronous lock in the async context
+    /// if it is guaranteed that the lock is not held for a long time.
+    verified_block_hashes: Arc<parking_lot::RwLock<BTreeSet<Hash256>>>,
+    validator_set: BTreeSet<PublicKey>,
+}
+
+impl MessageFilter for ConsensusMessageFilter {
+    fn filter(&self, message: &Message) -> Result<(), String> {
+        serde_json::from_str::<ConsensusMessage>(message.data()).map_err(|e| e.to_string())?;
+        if !self.validator_set.contains(message.signature().signer()) {
+            return Err("the signer is not in the validator set".to_string());
+        }
+        if self
+            .verified_block_hashes
+            .read()
+            .contains(&message.to_hash256())
+        {
+            Ok(())
+        } else {
+            Err("the block hash is not validated yet.".to_string())
+        }
```

Comment:
How could we validate a block even before receiving it?

Reply:
In addition, the error message in L78 says `not validated` while `verify` is used throughout the code.

Reply:
What do you mean by "verify"? Is it the same with validating a block?

Reply:
1. Block proposals IS NOT propagated via the consensus module (consensus module is only about their hashes). Proposals are handled by the Repository module, so we can expect that a block proposal is received in the repository first, and once verified, the node will register it to the consensus module.
2. I mistakenly use the two words "verify" and "validate" interchangeably. I believe all cryptographic things must be referred as 'verify' EXCEPT the word 'validator' which is used by an old convention. Fixed it.

Reply:
I thought the CSV module does not validate(word in blockchain context) blocks. Now I see the meaning of it, I think.